primitives/src/merkle_tree/mod.rs

```diff
@@ -0,0 +1,288 @@
+pub mod examples;
+pub mod macros;
+pub mod sparse_merkle_tree;
+
+mod internal;
+
+use crate::errors::PrimitivesError;
+use ark_std::{borrow::Borrow, fmt::Debug, string::ToString, vec, vec::Vec};
+use num_bigint::BigUint;
+use num_traits::ToPrimitive;
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug, Hash)]
+/// The result of querying at an index in the tree
+/// Typically, F for element type, P for membership proof type and N for
+/// non-membership proof type
+pub enum LookupResult<F, P, N> {
+    /// The value at the given index, and a proof of validity
+    Ok(F, P),
+    /// The index is valid but we do not have the leaf in memory
+    NotInMemory,
+    /// The index is outside the occupied range in the tree, and a
+    /// non-membership proof
+    EmptyLeaf(N),
+}
+
+impl<F, P, N> LookupResult<F, P, N> {
+    /// Assert the lookup result is Ok. Return a tuple of element and membership
+    /// proof.
+    pub fn expect_ok(self) -> Result<(F, P), PrimitivesError> {
+        match self {
+            LookupResult::Ok(x, proof) => Ok((x, proof)),
+            LookupResult::NotInMemory => Err(PrimitivesError::InternalError(
+                "Expected Ok, found NotInMemory".to_string(),
+            )),
+            LookupResult::EmptyLeaf(_) => Err(PrimitivesError::InternalError(
+                "Expected Ok, found EmptyLeaf".to_string(),
+            )),
+        }
+    }
+
+    /// Assert the lookup result is NotInMemory. Return a non-membership proof.
+    pub fn expect_empty(self) -> Result<N, PrimitivesError> {
+        match self {
+            LookupResult::EmptyLeaf(n) => Ok(n),
+            LookupResult::Ok(..) => Err(PrimitivesError::InternalError(
+                "Expected EmptyLeaf, found element".to_string(),
+            )),
+            LookupResult::NotInMemory => Err(PrimitivesError::InternalError(
+                "Expected EmptyLeaf, found NotInMemory".to_string(),
+            )),
+        }
+    }
+}
+
+/// An element of a Merkle tree.
+pub trait Element: Clone + Eq + PartialEq {}
+impl<T: Clone + Eq + PartialEq> Element for T {}
+
+/// An index type of a leaf in a Merkle tree.
+pub trait Index: Debug + Eq + PartialEq + Ord + PartialOrd + Clone + ToTraversalPath {}
+impl<T: Debug + Eq + PartialEq + Ord + PartialOrd + Clone + ToTraversalPath> Index for T {}
+
+/// An internal node value type in a Merkle tree.
+pub trait NodeValue: Default + Eq + PartialEq + Copy + Clone + Debug {}
+impl<T: Default + Eq + PartialEq + Copy + Clone + Debug> NodeValue for T {}
+
+/// Merkle tree hash function
+pub trait DigestAlgorithm<E, I, T>
+where
+    E: Element,
+    I: Index,
+    T: NodeValue,
+{
+    // Possible improvement: adding digest_element() and digest_index()
+    /// Digest a list of values
+    fn digest(data: &[T]) -> T;
+
+    /// Digest a leaf (an indexed element)
+    fn digest_leaf(pos: &I, elem: &E) -> T;
+}
+
+/// An trait for Merkle tree index type.
+pub trait ToTraversalPath {
```

Comment:
I wonder if this trait should be generic on arity, instead of making it an input of `to_traverse_path`. Fixing arity in the type system would allow optimizations like a specialized type where the path is determined by bit shifting, for 2-arity.